java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,153 @@
+  MDConstructor() {
+    exists(Method m | m = this.getMethod() |
+      m.getDeclaringType() instanceof MessageDigest and
+      m.hasName("getInstance")
+    )
+  }
+}
+
+/** The method `digest()` declared in `java.security.MessageDigest`. */
+class MDDigestMethod extends Method {
+  MDDigestMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("digest")
+  }
+}
+
+/** The method `update()` declared in `java.security.MessageDigest`. */
+class MDUpdateMethod extends Method {
+  MDUpdateMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("update")
+  }
+}
+
+/** The hashing method that could taint the input. */
+class MDHashMethodAccess extends MethodAccess {
+  MDHashMethodAccess() {
+    (
+      this.getMethod() instanceof MDDigestMethod or
+      this.getMethod() instanceof MDUpdateMethod
+    ) and
+    this.getNumArgument() != 0
+  }
+}
+
+/** Gets a regular expression for matching common names of variables that indicate the value being held is a password. */
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends Expr {
+  PasswordVarExpr() {
+    this.(VarAccess).getVariable().getName().toLowerCase().regexpMatch(getPasswordRegex()) and
+    not this.(VarAccess).getVariable().getName().toLowerCase().matches("%hash%") // Exclude variable names such as `passwordHash` since their values were already hashed
+  }
+}
+
+/** Holds if `Expr` e is a direct or indirect operand of `ae`. */
+predicate hasAddExpr(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
+
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
+predicate hasAnotherHashCall(MethodAccess ma) {
+  exists(MDHashMethodAccess ma2, DataFlow::Node node1, DataFlow::Node node2 |
+    ma2 != ma and
+    node1.asExpr() = ma.getAChildExpr() and
+    node2.asExpr() = ma2.getAChildExpr() and
+    (
+      TaintTracking2::localTaint(node1, node2) or
+      TaintTracking2::localTaint(node2, node1)
+    )
+  )
+}
+
+/** Holds if `MethodAccess` ma is a hashing call without a sibling node making another hashing call. */
+predicate isSingleHashMethodCall(MDHashMethodAccess ma) { not hasAnotherHashCall(ma) }
+
+/** Holds if `MethodAccess` ma is invoked by `MethodAccess` ma2 either directly or indirectly. */
+predicate hasParentCall(MethodAccess ma2, MethodAccess ma) {
+  ma.getCaller() = ma2.getMethod()
+  or
+  exists(MethodAccess ma3 |
+    ma.getCaller() = ma3.getMethod() and
+    hasParentCall(ma2, ma3)
+  )
+}
+
+/** Holds if `MethodAccess` is a single hashing call that is not invoked by a wrapper method. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not exists(MethodAccess ma2 | hasParentCall(ma2, ma)) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack to reduce FPs
```

Comment:
```suggestion
  not hasParentCall(_, ma)) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack. This reduces FPs.
```
This seems like a very significant carve-out -- we're excluding any hash-method call with a caller? I also think considering `ma2` is completely free you probably don't need the recursion in `hasParentCall` and are simply excluding all methods that we know of any call to (i.e., only methods that are exclusively called by external code unseen by us will be counted, such as `main` and request handlers called by an external library?)

Reply:
Updated